### PR TITLE
ForeignId RowData to add foreign id referring to a row from another t…

### PR DESCRIPTION
**Pull request by lemonboston**
…able. #62

---

Note: a copy of it is used already in opentaskspal

### CODE REVIEW

File: contentpal/src/main/java/org/dmfs/android/contentpal/rowdata/ForeignId.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * @author Gabor Keszthelyi
+ */
+public final class ForeignId<T> implements RowData<T>
+{
+    private final String mForeignIdColumnName;
+    private final RowSnapshot<?> mReferredRow;
+
+
+    public ForeignId(@NonNull String foreignIdColumnName, @NonNull RowSnapshot<?> referredRow)
```
Comment (dmfs) [inline contentpal/src/main/java/org/dmfs/android/contentpal/rowdata/ForeignId.java]: Please allow it to be used as a decorator as well. See CalendarPal `Located` for an example.

File: contentpal/src/main/java/org/dmfs/android/contentpal/rowdata/ForeignId.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import org.dmfs.android.contentpal.RowSnapshot;
+import org.dmfs.android.contentpal.TransactionContext;
+
+
+/**
+ * {@link RowData} that adds a foreign id referring to the provided row from another table.
+ *
+ * @author Gabor Keszthelyi
+ */
+public final class ForeignId<T> implements RowData<T>
```
Comment (dmfs) [inline contentpal/src/main/java/org/dmfs/android/contentpal/rowdata/ForeignId.java]: I think `Referring` is a better name as it expresses that this is `RowData` which adds a reference to another row.

**Review by dmfs — CHANGES_REQUESTED** [review]
Please also update the `Related` Operation to use this class and the `Populated` decorator instead.

**Comment by dmfs** [comment]
Maybe rename `Related` to `Referring` as well to point out the relation between these.

**Comment by lemonboston** [comment]
Made the suggested changes, ready for check again.

File: contentpal/src/main/java/org/dmfs/android/contentpal/operations/Referring.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+import android.support.annotation.NonNull;
+
+import org.dmfs.android.contentpal.InsertOperation;
+import org.dmfs.android.contentpal.Operation;
+import org.dmfs.android.contentpal.RowSnapshot;
+
+
+/**
+ * A {@link InsertOperation} decorator that a foreign key to another row of type {@code <V>}.
```
Comment (dmfs) [inline contentpal/src/main/java/org/dmfs/android/contentpal/operations/Referring.java]: Please also fix this Javadoc.

Comment (lemonboston) [reply contentpal/src/main/java/org/dmfs/android/contentpal/operations/Referring.java]: Fixed, also removed the generic type of the 'foreign table' - I suppose it's not needed.

File: calendarpal/src/main/java/org/dmfs/android/calendarpal/operations/EventRelated.java [hunk 4]
```diff
@@ -35,30 +31,12 @@
...
+public final class EventRelated<T> extends DelegatingOperation<T>
 {
 
-    private final Operation<T> mDelegate;
-
-
     public EventRelated(@NonNull RowSnapshot<CalendarContract.Events> event, @NonNull InsertOperation<T> delegate)
     {
-        mDelegate = new Related<>(event, "event_id"/* all supported tables use this name as the foreign key column name */, delegate);
+        super(new Referring<>(event, "event_id",  /* all supported tables use this name as the foreign key column name */ delegate));
```
Comment (dmfs) [inline calendarpal/src/main/java/org/dmfs/android/calendarpal/operations/EventRelated.java]: note the comment belongs to `"event_id"` not `delegate`, that's why the comma was to the right of the comment.

Comment (lemonboston) [reply calendarpal/src/main/java/org/dmfs/android/calendarpal/operations/EventRelated.java]: oh, I don't know how that comma crawled away..  fixed it

File: calendarpal/src/main/java/org/dmfs/android/calendarpal/operations/EventRelated.java [hunk 5]
```diff
@@ -35,30 +31,12 @@
  *
  * @author Marten Gajda
  */
-public final class EventRelated<T> implements InsertOperation<T>
+public final class EventRelated<T> extends DelegatingOperation<T>
```
Comment (dmfs) [inline calendarpal/src/main/java/org/dmfs/android/calendarpal/operations/EventRelated.java]: `EventRelated` should stay an `InsertOperation` even though the documentation doesn't explicitly forbid it, you really should not change the `event_id` of an attendee, reminder or extended data entry.

Comment (lemonboston) [reply calendarpal/src/main/java/org/dmfs/android/calendarpal/operations/EventRelated.java]: Reverted that change. Let me know if I should rebase and squash.

**Comment by dmfs** [comment]
please rebase and squash

**Comment by lemonboston** [comment]
Squashed and rebased.

**Review by dmfs — APPROVED** [review]
Looks good.

**Comment by dmfs** [comment]
Expect the code coverage dropped a bit ;-)

**Pull request merged**